reference/SDK: rework SDK docs,cover container SDK (#195)
flatcar-linux/flatcar-docs: docs/reference/developer-guides/sdk-modifying-flatcar.md

   **This is the recommended way of using the SDK.**

**NOTE** that currently, Docker is required to run the SDK.
While work on supporting other runtimes (e.g. Podman) is ongoing, the wrapper scripts currently only support Docker.

### Clone the scripts repo

The [scripts repository][scripts] - among other things - contains SDK wrapper scripts, a `version.txt` with release version information, and both the [coreos-overlay][coreos] and [portage-stable][portage] ebuild repositories as git submodules (more on ebuilds later).
A good way to think of the scripts repo is this being Flatcar's "SDK repo".

```shell
$ git clone https://github.com/flatcar-linux/scripts.git
$ cd scripts
$ git submodule init
$ git submodule update
```

#### Optionally, pick a release tag or branch

Cloning the repo will have it land on the `main` branch, which can be thought of as "alpha-next" - i.e. the next major Alpha release.
Even though main is smoke-tested in nightly builds, it might occasionally be broken in subtle ways.
This can make it harder to track down issues introduced by actual changes to Flatcar.

> * Release **tags** signify speficic (past) releases, like "stable-2905.2.4" or "beta-3033.1.1". Tags are created in release branches.

Reviewer: ```suggestion
* Release **tags** signify specific (past) releases, like "stable-2905.2.4" or "beta-3033.1.1". Tags are created in release branches.
```

After entering you're put right into the (host) script repository's bind mount root.
By default, the name of the container contains SDK and OS image version.
After starting, the version file will have been updated:

```shell
sdk@flatcar-sdk-all-3066_0_0_os-alpha-3066_0_0-5-gcf4ff44a ~/trunk/src/scripts $ cat sdk_container/.repo/manifests/version.txt
FLATCAR_VERSION=3066.0.0+5-gcf4ff44a
FLATCAR_VERSION_ID=3066.0.0
FLATCAR_BUILD_ID="5-gcf4ff44a"
FLATCAR_SDK_VERSION=3066.0.0
```

We're basing our work on release 3066.0.0 in this example, the current branch has 5 patches on top of that release, and the latest patch has the shortlog hash `cf4ff44a`.
This leads to `FLATCAR_BUILD_ID` being set (to the output of `git describe --tags`) and is reflected in the container name `...os-alpha-3066_0_0-5-gcf4ff44a`.

#### A note on persistence

`run_sdk_container` re-uses containers once started; containers to be re-used are identified by name (see above). 
Persistence helps with keeping changes in your work environment across container runs.
**Keep in mind though that a new container will be created if the working commit in the scripts repository changes**.

> This is usually derired to prevent version muddling.

Reviewer: ```suggestion
This is usually desired to prevent version muddling.
```

It can be explicitly overridden by using the `-n <name>` argument to `run_sdk_container`.

## Building an OS image

<table><tr><td>

**tl;dr** Build packages, base image, and vendor (qemu launchable) image.
This builds for the default architecture, `amd64-usr`.
Use `--board=arm64-usr` with packages / image script to build for ARM64.
```shell
sdk@flatcar-sdk $ ./build_packages
sdk@flatcar-sdk $ ./build_image
sdk@flatcar-sdk $ ./image_to_vm.sh <path-to-image--see-output-of-build_image>
```

</td></tr></table>

Before we discuss any modifications to the image, we'll do a full image build first. This will create a "known-good" base to mount your changes on.

### Select the target architecture

At the time of writing the SDK supports two target architectures: AMD64 (x86-64) and ARM64.
The target architecture can be specified by use of the `--board=` parameter to both `build_packages` and `build_image`:
* `--board=amd64-usr` will build an x86 image
* `--board=arm64-usr` will build and ARM64 image

> If no architecture is specified then AMD64 will be used by default.

Reviewer 2: Please, let's keep these two sentences about the distro package, the new NOTE below is too short and sounds like the user would have to just create this file.

We’ll cover everything you need to make the changes you want, and to produce an image for the runtime environment(s) you want to use Flatcar in (e.g. AWS, qemu, Packet, etc).
By the end of the guide you will build a developer image that you can run under qemu and have tools for making changes to the OS image like adding or removing packages, or shipping custom kernels.
Note that we chose this guide's "qemu" image target solely to enable local testing; the same process can be used to produce images for any and all targets (cloud providers etc.) supported by Flatcar.

**Note** there is a "tl;dr" paragraph at the start of each section which summarises the commands discussed in the section.

Flatcar Container Linux is an open source project. All of the source for Flatcar Container Linux is available on [github][github-flatcar]. If you find issues with these docs or the code please send a pull request.

Please direct questions and suggestions to the [#flatcar:matrix.org Matrix channel][matrix] or [mailing list][flatcar-dev].

## Getting started

<table><tr><td>

**tl;dr** Check out a release branch and start the SDK (this uses the current Alpha release branch).
```shell

> $ git clone https://github.com/flatcar-linux/scripts.git

Reviewer 2: ```suggestion
$ git clone --recurse-submodules https://github.com/flatcar-linux/scripts.git
```
I would always use this in the docs for people to be aware that submodules are needed (even if we have automation to fetch them when they are missing)

          |                  +------portage-stable
          `---------.repo
                     +----manifests
                           +-------- version.txt
```

There are 2 ways to use the SDK container:
1. Standalone: Run the container and clone the scripts repo inside the container.
   This is great for one-shot SDK usage; it's not optimal for sustained OS development since versioning is unclear and changes might get lost.
2. Wrapped: Uses a wrapper script to run the container and to bind-mount the local scripts directory into the container.
   **This is the recommended way of using the SDK.**

**NOTE** The SDK container supports being run on docker or podman, with docker taking preference when both are available.
The wrapper scripts will auto-detect which one is available, and use it.

### Clone the scripts repo

The [scripts repository][scripts] - among other things - contains SDK wrapper scripts, a `version.txt` with release version information, and both the [coreos-overlay][coreos] and [portage-stable][portage] ebuild repositories as git submodules (more on ebuilds later).
A good way to think of the scripts repo is this being Flatcar's "SDK repo".

```shell

> $ git clone https://github.com/flatcar-linux/scripts.git

Reviewer 2: ```suggestion
$ git clone --recurse-submodules https://github.com/flatcar-linux/scripts.git
```

It is generally recommended to base work on the latest Alpha release.
While new features should target `main` at merge time, Alpha is a tested release and therefore offers a more stable foundation to base work on.
At the same time, Alpha is not too far away from `main` so the risk of merge-time conflicts should be low.

Find the latest Alpha release branch:

```shell
$ git branch -r -l | sed -n 's:origin/\(flatcar-[0-9]\+\)$:\1:p' | sort | tail -n1
```

If the goal is to reproduce and to fix a bug of a release other than Alpha, it is recommended to base the work on the latest point release of the respective major version instead of Alpha. All currrently "active" major versions can be found at the top of the [releases][flatcar-releases] web page.

For quick reference, to get the latest stable release tag, use:
```shell
$ git tag -l | grep -E 'stable-[0-9.]+$' | sort | tail -n 1
```
(replace `stable` with `beta` or `alpha` in accordance with your needs).

We provide a helper script to check out tags or branches and keeping submodules aligned. Just run:

```shell
$ ./checkout [branch-or-tag-from-above]
```

> This wikk check out the branch or tag in the top-level `scripts` directory and update the submodules, depending on wheter a branch or tag was specified:

Reviewer 2: ```suggestion
This will check out the branch or tag in the top-level `scripts` directory and update the submodules, depending on wheter a branch or tag was specified:
```

**Note** there is a "tl;dr" paragraph at the start of each section which summarises the commands discussed in the section.

Flatcar Container Linux is an open source project. All of the source for Flatcar Container Linux is available on [github][github-flatcar]. If you find issues with these docs or the code please send a pull request.

Please direct questions and suggestions to the [#flatcar:matrix.org Matrix channel][matrix] or [mailing list][flatcar-dev].

## Getting started

<table><tr><td>

**tl;dr** Check out a release branch and start the SDK (this uses the current Alpha release branch).
```shell
$ git clone https://github.com/flatcar-linux/scripts.git
$ cd scripts
$ branch="$(git branch -r -l | sed -n 's:origin/\(flatcar-[0-9]\+\)$:\1:p' | sort | tail -n1)"
$ git checkout "$branch"
$ git submodule init
$ git submodule update
$ for s in sdk_container/src/third_party/coreos-overlay/ sdk_container/src/third_party/portage-stable; do
$   git -C "$s" checkout "$branch"
$   git -C "$s" pull --rebase
$ done

> $ ./run_sdk_container -t

Reviewer 2: This fails if it was run once and the `version.txt` got created:
```
Error: error creating container storage: the container name "flatcar-sdk-all-3087.0.0_os-alpha-3066.0.0-52-g1933bf20" is already in use
```

Reviewer 2: adding `-n NAME` as mentioned later in the docs gives the same error

Reviewer 2: ah, it's the `^/` not working with podman...

Author: Interesting - could you please `set -x` at the top of `run_sdk_container` and provide the output? It works here...
(Also - is that with Docker or with Podman?)

Reviewer 2: It was with Podman, pushed a fix in https://github.com/flatcar-linux/scripts/pull/199 by leaving the `^/` prefix out

Author: Good catch - most interestingly I just added the regex today to fix something else: https://github.com/flatcar-linux/scripts/pull/196

This is usually desired to prevent version muddling.
It can be explicitly overridden by using the `-n <name>` argument to `run_sdk_container`.

## Building an OS image

<table><tr><td>

**tl;dr** Build packages, base image, and vendor (qemu launchable) image.
This builds for the default architecture, `amd64-usr`.
Use `--board=arm64-usr` with packages / image script to build for ARM64.
```shell
sdk@flatcar-sdk $ ./build_packages
sdk@flatcar-sdk $ ./build_image
sdk@flatcar-sdk $ ./image_to_vm.sh <path-to-image--see-output-of-build_image>
```

</td></tr></table>

Before we discuss any modifications to the image, we'll do a full image build first. This will create a "known-good" base to mount your changes on.

### Select the target architecture

**NOTE on cross-compilation**: if you are cross-compiling make sure a static aarch64 qemu is set up via binfmt-misc on your host machine.
Some packages compile and execute intermediate commands during their build process - this can break cross-compiling since the commands are built for the target architecture.
The qemu binary on the host needs to be a static binary since it will be called from within the container context.

> Check if your distro has support for aarch64 in `binfmt-misc` already; on e.g. Fedora there's an `qemu-aarch64` entry in `/proc` for that (the name of the proc file may vary across distributions though):

Reviewer 2: ```suggestion
Check if your distro has a `qemu-user-static` package that you can install or whether it has support for aarch64 in `binfmt-misc` already; on e.g. Fedora there's an `qemu-aarch64` entry in `/proc` for that (the name of the proc file may vary across distributions though):
```

You can run `docker run --rm -ti arm64v8/alpine` on your host system as an easy check to verify everything is ready.

### Build the OS image packages

It's likely this won't *actually* build the packages but rather download pre-built packages from the Flatcar binary package cache.
The package cache is updated on every release.

```shell
$ ./build_packages [--board=...]
```

> The command should download most packages from our binary cache - speeding up the "build" - since we are basing this on an existing release.

Reviewer 2: I think this needs more info on how to make sure that it is set up. In the tips and tricks all the `set_version` commands are removed and the user is left without any info on how to set the package cache up, be it for the maintenance branch or the main branch.

Author: They should not need to since information on the binary caches is included in the SDK container, which itself is versioned. `set_version` disappeared because it is replaced (implicitly) by checking out the correct branch or tag of the scripts repo. 

It works like this:
1. check out version A, update submodules
2. `./run_sdk_container` will read the versionfile, use the correct SDK container image for version A
3. `./build_packages` in the container will use the correct bincache for version A.

Now, to switch versions, one would:
1. check out version B, update submodules
2. `./run_sdk_container` now uses a different SDK container image - for version B
3. `./build_packages` in the container will use the bincache for version B.

Additionally, the containers for version A and B will be named accordingly (the container name contains OS and SDK version) - so one could switch back and forth between versions by just checking out the respective `scripts` branch, and never lose intermediate build artifacts (as long as the container is not deleted by the user).

Author: Is there something I'm missing that `set_version` does which couldn't be done by using the plain versionfile from `scripts`?

Reviewer 2: Ok, so the idea would be to publish nightly SDK containers for the main and the maintenance branches? Maybe without building the SDK since it doesn't need to be rebuilt for maintenance branches?

Author: Yes, that would be the idea. The containerised nightly job for main already builds the SDK container and credentials for pushing from CI to ghcr.io are in place (but currently unused) - it's a one-line change to publish after building.

Maintenance (release) branches don't need a new SDK (made a mistake in my first comment and updated it); the versionfile should determine the bincache URL.

Reviewer 2: While this is not done I suggest to keep the `set_version` docs around for getting binary packages when working on the main/maintenance branches

Reviewer 2: > the other [solution] is that we can use the new OS version to overwrite the contents in `make.conf`

Thinking again this (overwriting the portage make.conf binhost when the OS version file gets created) is the only real solution because the current behavior is not save for releases either - they could end up reusing the initial major version binary packages instead of rebuilding all of them (didn't check if this is already the case on the nightly test job in jenkins).

Author: > Thinking again this (overwriting the portage make.conf binhost when the OS version file gets created) is the only real solution because the current behavior is not save for releases either - they could end up reusing the initial major version binary packages instead of rebuilding all of them (didn't check if this is already the case on the nightly test job in jenkins).

That's a great catch - I'll have a look at adding this to [sdk_entry.sh](https://github.com/flatcar-linux/scripts/blob/main/sdk_lib/sdk_entry.sh) so the container's `make.conf` is always aligned with the versionfile.

Author: @pothos https://github.com/flatcar-linux/scripts/pull/201

Reviewer 2: With https://github.com/flatcar-linux/scripts/pull/209 I think I should be happy with the reached state and we can resolve this, maybe it's good to tell people how to get the latest nightly `version.txt` from a tag with `git show REF:sdk_container/…/version.txt` until we push it each night to the maintenance/main branch

Author: Good point, I'll make a mental note to update the SDK docs with instructions how to use the nightly tags after flatcar-linux/scripts#209 is merged and our CI is updated.

You can run `docker run --rm -ti arm64v8/alpine` on your host system as an easy check to verify everything is ready.

At the time of writing the SDK supports two target architectures: AMD64 (x86-64) and ARM64.
The target architecture can be specified by use of the `--board=` parameter to both `build_packages` and `build_image`:
* `--board=amd64-usr` will build an x86 image
* `--board=arm64-usr` will build and ARM64 image

If no architecture is specified then AMD64 will be used by default.

### Build the OS image packages

> It's likely this won't *actually* build the packages but rather download pre-built packages from the Flatcar binary package cache.

Reviewer 2: ```suggestion
Beware, it's likely this won't *actually* build the packages but rather download pre-built packages from the Flatcar binary package cache (see below on how to force-rebuild a package that you modified).
```

For other vendor images, pass the `--format=` parameter (see `./image_to_vm.sh --help`).
In general, `image_to_vm.sh` will read the generic disk image, install any vendor specific tools to the OEM partition where applicable (e.g. Azure VM tools for the Azure VM), and produce a vendor specific image. In the case of QEMU, a qcow2 image is produced. QEMU does not require vendor specific tooling in the OEM partition.

> On the host outside the container, the image(s) built are located in `__build__/images/...`.

Reviewer: ```suggestion
On the host outside the container, the image(s) built are located in `__build__/images/…`.
```

Now for the interesting part! We are going to discuss 2 ways of making changes: adding or upgrading a package, and modifying the kernel configuration.

> ### A brief introduction to Gentoo how it relates to the SDK

Reviewer: I think an "and" is missing here:

```suggestion
### A brief introduction to Gentoo and how it relates to the SDK
```

Flatcar Container Linux is based on ChromiumOS, which is based on Gentoo.
While the ChromiumOS heritage has faded and is barely visible nowadays, we heavily leverage Gentoo processes and tools.

Contrary to traditional Linux distributions, Gentoo applications and “packages” are compiled at installation time.
Gentoo itself does not ship packages - instead, it consists of a massive number of ebuild files to build applications at installation time (that’s an oversimplification as there are binary package caches, but that’s beyond the scope of this document).

> While the Flatcar SDK can be understood as a Gentoo derivate, the OS image is special.

Reviewer: ```suggestion
While the Flatcar SDK can be understood as a Gentoo derivative, the OS image is special.
```

Each package directory may contain ebuild files for multiple versions, e.g. `dev-lang/python` contains a host of python versions (used in the SDK).
Furthermore, each package directory contains a `Manifest` file with cryptographic checksums and file sizes of the package's source tarball(s), and may contain a `files/` directory containing auxiliary files to build / install the package, e.g. patches or config files.

Multiple package sources - in separate directories - can be stacked on top of each other.
These “overlays” allow custom extensions or even custom sub-trees on top of an existing foundation.
In these stacks, “upper” level packages override “lower” level ones.
The Flatcar build system uses a fork of Gentoo upstream’s [portage-stable](https://github.com/flatcar-linux/portage-stable) as its base, and the overlay repository [coreos-overlay](https://github.com/flatcar-linux/coreos-overlay) for Flatcar specific modifications and packages on top.

Packages are built using "ebuild" files.

> These files contain a package’s dependencies - both build and runtime - as well as implement callbacks for downloading, patching, building, and installing a package.

Reviewer: ```suggestion
These files contain package’s dependencies - both build and runtime - as well as implement callbacks for downloading, patching, building, and installing a package.
```

Reviewer: I'd say that the "a" article is bound to the "dependencies" word (and since "dependencies" is plural, it shouldn't have any indefinite article), not to "package's". But not an expert on articles. Anyway, I wouldn't go with "packages' dependencies", because the sentence is about a single package. So how about:

```suggestion
These files contain dependencies of a package - both build and runtime - as well as implement callbacks for downloading, patching, building, and installing the package.
```

The OS image packages therefore have their own root inside the SDK - AMD64 is located at `/build/amd64-usr/` and ARM64 is under `/build/arm64-usr/`.

Both board chroot and SDK use Gentoo's portage to manage its respective packages: `sudo emerge` is used to manage SDK packages, and `emerge-<arch>` (`emerge-amd64-usr` or `emerge-arm64-usr`, without sudo) is used to do the same for the OS image roots.

## Add (or update) a package

All of the following is done inside the SDK container, i.e. after running
```shell
$ ./run_sdk_container.sh -t
```

<table><tr><td>

**tl;dr** In the SDK container, introduce a new upstream package from Gentoo.
```shell
~/trunk/src/scripts $ git clone --depth 5 https://github.com/gentoo/gentoo.git
~/trunk/src/scripts $ mkdir -p ../third_party/portage-stable/<group>/
~/trunk/src/scripts $ cp -R gentoo/<group>/<package> ../third_party/portage-stable/<group>/
~/trunk/src/scripts $ emerge-amd64-usr --newuse <group>/<package>
# optional - add missing eclass
~/trunk/src/scripts $ cp gentoo/eclass/<eclass-name>.eclass ../third_party/portage-stable/eclass/
~/trunk/src/scripts $ emerge-amd64-usr --newuse <group>/<package>

> # optional - unmask package
> ~/trunk/src/scripts $ vim ../third_party/portage-stable/<group>/<package>/<package-version>.ebuild

Reviewer: I'd say that this is a "no no". Instead, the preference is to modify the `package.accept_keywords` in coreos-overlay, so the modification won't be lost when we sync the portage-stable with gentoo next time. So maybe:
```suggestion
# optional - unmask package
~/trunk/src/scripts $ vim ../third_party/coreos-overlay/profiles/coreos/base/package.accept_keywords
```

Author: Agreed, makes sense - in particular since we're striving for full alignment and potential deprecation of portage-stable.

Packages are built using "ebuild" files.
These files contain dependencies of a package - both build and runtime - as well as implement callbacks for downloading, patching, building, and installing the package.
The callbacks in these ebuild files are written in shell.
The Gentoo package system - portage - will, when building / installing a package, run the respective callbacks in order (e.g. `src_fetch()` for downloading package sources, and `src_compile()` for building).
Common ebuild functions shared across many packages are implemented via classes (in `eclass/`) which can be inherited by package ebuilds.

For more information on Gentoo in general please refer to the [Gentoo devmanual](https://devmanual.gentoo.org/).

### Get to know the SDK chroot

When entering the SDK you are in the `~/trunk/src/scripts` repository which can be seen as the build system.
It is one of the three repositories that define a Flatcar build:

> 1. flatcar-scripts (the directory you're in) contains high-level build scripts to build all packages for an image, to build an image, and to bootstrap an SDK.

Reviewer 2: ```suggestion
1. `scripts` (the directory you're in) contains high-level build scripts to build all packages for an image, to build an image, and to bootstrap an SDK.
```

## Add (or update) a package

All of the following is done inside the SDK container, i.e. after running
```shell
$ ./run_sdk_container.sh -t
```

<table><tr><td>

**tl;dr** In the SDK container, introduce a new upstream package from Gentoo.
```shell
~/trunk/src/scripts $ git clone --depth 5 https://github.com/gentoo/gentoo.git
~/trunk/src/scripts $ mkdir -p ../third_party/portage-stable/<group>/
~/trunk/src/scripts $ cp -R gentoo/<group>/<package> ../third_party/portage-stable/<group>/
~/trunk/src/scripts $ emerge-amd64-usr --newuse <group>/<package>
# optional - add missing eclass
~/trunk/src/scripts $ cp gentoo/eclass/<eclass-name>.eclass ../third_party/portage-stable/eclass/
~/trunk/src/scripts $ emerge-amd64-usr --newuse <group>/<package>
# optional - unmask package
~/trunk/src/scripts $ vim ../third_party/coreos-overlay/profiles/coreos/base/package.accept_keywords
# remove '~' from arm64 and amd64
~/trunk/src/scripts $ emerge-amd64-usr --newuse <group>/<package>
# optional - add missing dependencies, see line 2 ff. above
~/trunk/src/scripts $ ./build_image
~/trunk/src/scripts $ ./image_to_vm.sh --from=../build/images/amd64-usr/latest --format qemu

> ~/trunk/src/scripts $ ../build/images/amd64-usr/latest/flatcar_production_qemu.sh &

Reviewer 2: The qemu scripts may write to the terminal, I would not use `&` here but rather recommend the user to open a new terminal tab for SSH outside of this SDK container.

Reviewer 2: Actually I would do both outside the SDK container since it's easier to pass an Ignition config to the qemu script this way.

Multiple package sources - in separate directories - can be stacked on top of each other.
These “overlays” allow custom extensions or even custom sub-trees on top of an existing foundation.
In these stacks, “upper” level packages override “lower” level ones.
The Flatcar build system uses a fork of Gentoo upstream’s [portage-stable](https://github.com/flatcar-linux/portage-stable) as its base, and the overlay repository [coreos-overlay](https://github.com/flatcar-linux/coreos-overlay) for Flatcar specific modifications and packages on top.

Packages are built using "ebuild" files.
These files contain dependencies of a package - both build and runtime - as well as implement callbacks for downloading, patching, building, and installing the package.
The callbacks in these ebuild files are written in shell.
The Gentoo package system - portage - will, when building / installing a package, run the respective callbacks in order (e.g. `src_fetch()` for downloading package sources, and `src_compile()` for building).

> Common ebuild functions shared across many packages are implemented via classes (in `eclass/`) which can be inherited by package ebuilds.

Reviewer: ```suggestion
Common ebuild functions shared across many packages are implemented via eclasses (in `eclass/`) which can be inherited by package ebuilds.
```

  # required by =<group>/package> (argument)
  =<other-group>/<other-package> **
```

> To proceed, remove the architecture mask in the ebuild file (remove the `~` - for example, change `~arm64` to `arm64`) and re-run emerge.

Reviewer: In the same vein as above:
```suggestion
To proceed, add the package name and version, and its masked architectures to the `package.accept_keywords` file inside the `coreos` profile. Which `package.accept_keywords` file should be updated depends on couple factors - whether it is needed for both SDK and OS image or only for SDK or only for OS image, whether it is needed for both AMD64 and ARM64 images, or only for AMD64 or only for ARM64. Please refer to `README.md` in coreos-overlay for a summary about profiles
```

> FLatcar follows its own stabilisation process (through the Alpha - Beta - Stable channels); it's perfectly fine to unmask a package upstream considers unstable.

Reviewer: ```suggestion
Flatcar follows its own stabilisation process (through the Alpha - Beta - Stable channels); it's perfectly fine to unmask a package upstream considers unstable.
```

~/trunk/src/scripts $ cd ~/trunk/src/scripts 
~/trunk/src/scripts $ emerge-amd64-usr sys-kernel/coreos-kernel
~/trunk/src/scripts $ emerge-amd64-usr sys-kernel/coreos-modules
~/trunk/src/scripts $ ./build_image --board=amd64-usr
~/trunk/src/scripts $ ./image_to_vm.sh --from=../build/images/amd64-usr/latest --board=amd64-usr --format qemu
~/trunk/src/scripts $ ../build/images/amd64-usr/latest/flatcar_production_qemu.sh
~/trunk/src/scripts $ ssh core@localhost -p 2222
```

</td></tr></table>

Next, we’ll look into changing the kernel configuration - e.g. for adding a kernel module or a core kernel feature not shipped with stock Flatcar.
This will give you a deep dive into the low-level bits of Gentoo's build and packaging system.
To modify the configuration of a package we will run its individual build steps manually - by use of `ebuild` instead of `emerge`.
This will allow for pausing after downloading the sources, to change the source tree configuration before building and installing.

Our first step is to set you all up with a pre-configured stock Flatcar Linux kernel to base your modifications on.

> The Flatcar Linux kernel build is split over multiple gentoo ebuild files which all reside in <code>[coreos-overlay/sys-kernel/](https://github.com/kinvolk/coreos-overlay/tree/main/sys-kernel)</code>:

Reviewer: Won't this work?

```suggestion
The Flatcar Linux kernel build is split over multiple gentoo ebuild files which all reside in [`coreos-overlay/sys-kernel/`](https://github.com/kinvolk/coreos-overlay/tree/main/sys-kernel):
```

  # required by =<group>/package> (argument)
  =<other-group>/<other-package> **
```

> To proceed, add the package name and version, and its masked architectures to the `package.accept_keywords` file inside the `coreos` profile. Which `package.accept_keywords` file should be updated depends on couple factors - whether it is needed for both SDK and OS image or only for SDK or only for OS image, whether it is needed for both AMD64 and ARM64 images, or only for AMD64 or only for ARM64. Please refer to `README.md` in coreos-overlay for a summary about profiles

Reviewer: ```suggestion
To proceed, add the package name and version, and its masked architectures to the `package.accept_keywords` file inside the `coreos` profile. Which `package.accept_keywords` file should be updated depends on couple factors - whether it is needed for both SDK and OS image or only for SDK or only for OS image, whether it is needed for both AMD64 and ARM64 images, or only for AMD64 or only for ARM64. Please refer to `README.md` in coreos-overlay for a summary about profiles.
```
